File: sdks/java/src/main/java/com/featuresignals/sdk/FeatureSignalsProvider.java
package com.featuresignals.sdk;

import java.util.Map;

/**
 * OpenFeature-compatible provider backed by FeatureSignalsClient.
 *
 * <p>Implements the resolution methods that the OpenFeature Java SDK expects.
 * Can be used directly without the OpenFeature dependency.
 */
public class FeatureSignalsProvider implements AutoCloseable {
    private final FeatureSignalsClient client;

    public FeatureSignalsProvider(String sdkKey, ClientOptions options) {
        this.client = new FeatureSignalsClient(sdkKey, options);
    }

    public FeatureSignalsClient getClient() { return client; }
    public String getName() { return "featuresignals"; }

    public ResolutionDetails<Boolean> resolveBooleanEvaluation(String flagKey, boolean defaultValue) {
        return resolve(flagKey, defaultValue, Boolean.class);
    }

    public ResolutionDetails<String> resolveStringEvaluation(String flagKey, String defaultValue) {
        return resolve(flagKey, defaultValue, String.class);
    }

    public ResolutionDetails<Double> resolveNumberEvaluation(String flagKey, double defaultValue) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null) {
            return new ResolutionDetails<>(defaultValue, "ERROR", "FLAG_NOT_FOUND");
        }
        if (val instanceof Number) {
            return new ResolutionDetails<>(((Number) val).doubleValue(), "CACHED", null);
        }
        return new ResolutionDetails<>(defaultValue, "ERROR", "TYPE_MISMATCH");
    }

    public ResolutionDetails<Object> resolveObjectEvaluation(String flagKey, Object defaultValue) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null) {
            return new ResolutionDetails<>(defaultValue, "ERROR", "FLAG_NOT_FOUND");
        }
        return new ResolutionDetails<>(val, "CACHED", null);
    }

    @Override
    public void close() {
        client.close();
    }

    @SuppressWarnings("unchecked")
    private <T> ResolutionDetails<T> resolve(String flagKey, T defaultValue, Class<T> type) {
        Map<String, Object> flags = client.allFlags();
        Object val = flags.get(flagKey);
        if (val == null) {
            return new ResolutionDetails<>(defaultValue, "ERROR", "FLAG_NOT_FOUND");
        }
        if (type.isInstance(val)) {
            return new ResolutionDetails<>((T) val, "CACHED", null);
        }
        return new ResolutionDetails<>(defaultValue, "ERROR", "TYPE_MISMATCH");
    }

    public record ResolutionDetails<T>(T value, String reason, String errorCode) {}
}

File: sdks/dotnet/src/FeatureSignals/OpenFeature/FeatureSignalsProvider.cs
using System.Text.Json;

namespace FeatureSignals.OpenFeature;

/// <summary>
/// OpenFeature-compatible provider backed by <see cref="FeatureSignalsClient"/>.
/// </summary>
public sealed class FeatureSignalsProvider : IDisposable
{
    private readonly FeatureSignalsClient _client;

    public FeatureSignalsProvider(string sdkKey, ClientOptions options)
    {
        _client = new FeatureSignalsClient(sdkKey, options);
    }

    public string Name => "featuresignals";

    public FeatureSignalsClient Client => _client;

    public ResolutionDetails<bool> ResolveBooleanEvaluation(
        string flagKey, bool defaultValue, EvalContext? context = null)
    {
        return Resolve(flagKey, defaultValue, val =>
            val is true or false ? (bool)val : throw new InvalidCastException());
    }

    public ResolutionDetails<string> ResolveStringEvaluation(
        string flagKey, string defaultValue, EvalContext? context = null)
    {
        return Resolve(flagKey, defaultValue, val =>
            val is string s ? s : throw new InvalidCastException());
    }

    public ResolutionDetails<double> ResolveNumberEvaluation(
        string flagKey, double defaultValue, EvalContext? context = null)
    {
        return Resolve(flagKey, defaultValue, val => val switch
        {
            long l => l,
            double d => d,
            int i => i,
            _ => throw new InvalidCastException()
        });
    }

    public ResolutionDetails<T> ResolveObjectEvaluation<T>(
        string flagKey, T defaultValue, EvalContext? context = null)
    {
        var flags = _client.AllFlags();
        if (!flags.TryGetValue(flagKey, out var raw))
        {
            return new ResolutionDetails<T>
            {
                Value = defaultValue,
                ErrorCode = ErrorCode.FlagNotFound,
                ErrorMessage = $"flag '{flagKey}' not found"
            };
        }

        try
        {
            if (raw is T typed)
                return new ResolutionDetails<T> { Value = typed };

            var json = JsonSerializer.Serialize(raw);
            var deserialized = JsonSerializer.Deserialize<T>(json);
            return new ResolutionDetails<T> { Value = deserialized! };
        }
        catch
        {
            return new ResolutionDetails<T>
            {
                Value = defaultValue,
                ErrorCode = ErrorCode.TypeMismatch,
                ErrorMessage = $"cannot convert to {typeof(T).Name}"
            };
        }
    }

    public void Shutdown() => _client.Dispose();

    public void Dispose() => Shutdown();

    private ResolutionDetails<T> Resolve<T>(
        string flagKey, T defaultValue, Func<object?, T> cast)
    {
        var flags = _client.AllFlags();
        if (!flags.TryGetValue(flagKey, out var raw))
        {
            return new ResolutionDetails<T>
            {
                Value = defaultValue,
                ErrorCode = ErrorCode.FlagNotFound,
                ErrorMessage = $"flag '{flagKey}' not found"
            };
        }

        try
        {
            return new ResolutionDetails<T> { Value = cast(raw) };
        }
        catch
        {
            return new ResolutionDetails<T>
            {
                Value = defaultValue,
                ErrorCode = ErrorCode.TypeMismatch,
                ErrorMessage = $"expected {typeof(T).Name}, got {raw?.GetType().Name ?? "null"}"
            };
        }
    }
}
